kernel/src/expressions/column_names.rs

```diff
@@ -106,6 +129,131 @@ impl Hash for ColumnName {
     }
 }
 
+/// A wrapper for `ColumnName` that pretty-prints the column name with fields delimited by periods
+/// and enclosed in square brackets, e.g. `column_name!("a.b.c")` would print as `"[a].[b].[c]"`.
+// TODO: Marked as test-only for now, as a demonstration. Should this be Display for ColumnName, so
+// `ColumnName::new(["a", "b.c", "d"])` would print as `"a.[b.c].d"` instead of merely `"a.b.c.d"`?
+#[cfg(test)]
+struct FancyColumnName(ColumnName);
+
+#[cfg(test)]
+impl Display for FancyColumnName {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut delim = None;
+        for s in self.0.iter() {
+            use std::fmt::Write as _;
+
+            // Only emit the period delimiter after the first iteration
+            if let Some(d) = delim {
+                f.write_char(d)?;
+            } else {
+                delim = Some('.');
+            }
+
+            if s.is_empty() || s.contains(|c: char| !c.is_ascii_alphanumeric() && c != '_') {
+                // Special situation detected. For safety, surround the field name with brackets
+                // (with proper escaping if the field name itself contains brackets).
+                f.write_char('[')?;
+                for c in s.chars() {
+                    match c {
+                        ']' => f.write_str("]]")?,
+                        _ => f.write_char(c)?,
+                    }
+                }
+                f.write_char(']')?;
+            } else {
+                // The fild name contains no special characters, so emit it as-is.
+                f.write_str(s)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+use crate::error::{DeltaResult, Error};
+
+/// Parses a column name from a string. Field names are separated by dots. Field names enclosed in
+/// square brackets may contain arbitrary characters, including periods and spaces. To include a
+/// literal square bracket in a field name, escape it by doubling it, e.g. `"[foo]]bar]"` would
+/// parse as `foo]bar`.
+// TODO: Marked as test-only for now, as a demonstration. Should we officially support it?
+#[cfg(test)]
+impl std::str::FromStr for ColumnName {
+    type Err = Error;
+
+    fn from_str(s: &str) -> DeltaResult<Self> {
+        column_name_from_str(s)
+    }
+}
+
+/// See `impl FromStr for ColumnName`.
+#[cfg(test)]
+fn column_name_from_str(s: &str) -> DeltaResult<ColumnName> {
+    type Chars<'a> = std::iter::Peekable<std::str::Chars<'a>>;
+
+    fn parse_simple(chars: &mut Chars<'_>) -> DeltaResult<(String, bool)> {
+        let mut name = String::new();
+        let mut allow_digits = false; // first character cannot be a digit
+        for c in chars {
+            match c {
+                '_' | 'a'..='z' | 'A'..='Z' => name.push(c),
+                '0'..='9' if allow_digits => name.push(c),
```

Comment:
nit: if something starts with a digit say `2` the error will be:
`"Invalid character '2' in unescaped field name"`

which is a little confusing. Since we're building an error string here we could just do:
```suggestion
                '0'..='9' => {
                  if allow_digits {
                    name.push(c)
                  } else {
                    return Err(Error::generic(format!(
                      "Field name cannot start with digit: {c}"
                    )))
                  }
                }
                   
```